wasm_runtime/src/execution_ctx.rs

```diff
@@ -207,6 +197,51 @@ impl WasmExecutionCtx {
 
         stdout_buf.clone()
     }
+
+    // Helper function to insert a Wasm execution context from the HashMap
+    //  
+    // It checks for duplicated `executionctx_id`.
+    // Returns Result<String, String>, with the inserted Wasm execution context id,
+    // otherswise, in case of error the String will contain the reason.
+    //
+    fn try_insert(wasm_executionctx: WasmExecutionCtx) -> Result<String, String> {
+        let executionctx_id = wasm_executionctx.id.clone();
+
+        // get write access to the WasmExecutionCtx HashMap
+        let mut executionctxs = WASM_RUNTIME_EXECUTIONCTXS.write()
+            .expect("ERROR! Poisoned RwLock WASM_RUNTIME_EXECUTIONCTXS on write()");
```

Comment:
With `expect` we will be panicking here, I think we want to return an error instead. Following the current API, something like:

```suggestion
let mut executionctxs = WASM_RUNTIME_EXECUTIONCTXS.write().map_err(|e| format!("error: {:?}", e))?;
```

I would do something like that in the rest of the API where we want to return an error, instead of panicking.

As a last comment, the problem with our type `Result<String, String>` is that for consumers, it's very hard to understand what went wrong if something did, they would have to "parse" the error part of the response. Instead, I think we should define something like `Result<String, ErrorType>`, where `ErrorType` is an enum with the possible errors.

I have to admit that mod_wasm is a bit of a special animal, because these "rich" types will be handled in C, and then somehow reported to Apache, so we might want to keep Strings on the errors if we are forwarding them to C and just showing them to the user and not doing any kind of special logic.

Reply:
Regarding error types, I agree 100%. Indeed, we mentioned that long time ago and created #4 with that purpose 😄

Reply:
Also, created #43 to deal with poisoned locks.